File: src/main/java/com/example/gava/controller/TodoController.java
package com.example.gava.controller;

import com.example.gava.entity.Todo;
import com.example.gava.service.TodoService;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.time.LocalDate;
import java.time.LocalDateTime;

@RestController
@RequestMapping("/todos")
public class TodoController {
    private final TodoService todoService;

    public TodoController(TodoService todoService) {
        this.todoService = todoService;
    }

    @GetMapping
    public Flux<Todo> getAllTodos() {
        return todoService.getAllTodos();
    }

    @GetMapping("/sortedByPriority")
    public Flux<Todo> getAllTodosSortedByPriority() {
        return todoService.getAllTodosSortedByPriority();
    }

    @GetMapping("/sortedByDueDate")
    public Flux<Todo> getAllTodosSortedByDueDate() {
        return todoService.getAllTodosSortedByDueDate();
    }

    @GetMapping("/{id}")
    public Mono<Todo> getTodoById(@PathVariable Long id) {
        return todoService.getTodoById(id);
    }

    @PostMapping
    public Mono<Todo> createOrUpdate(@RequestBody Todo todo) {
        return todoService.createOrUpdate(todo);
    }

    @DeleteMapping("/{id}")
    public Mono<Void> deleteTodoById(@PathVariable Long id) {
        return todoService.deleteTodoById(id);
    }

    @GetMapping("/byDate")
    public Flux<Todo> getTodosByDate(@RequestParam("date") LocalDate date) {
        return todoService.getTodosByDate(date);
    }

    @GetMapping("/byMonth")
    public Flux<Todo> getTodosByMonth(@RequestParam("year") int year, @RequestParam("month") int month) {
        return todoService.getTodosByMonth(year, month);
    }

    @GetMapping("/byYear")
    public Flux<Todo> getTodosByYear(@RequestParam("year") int year) {
        return todoService.getTodosByYear(year);
    }

    @GetMapping("/api/todos/completed/count")
    public Mono<Long> getCompletedTodoCount(
            @RequestParam("startDate") LocalDateTime startDate,
            @RequestParam("endDate") LocalDateTime endDate) {
        return todoService.getCompletedTodoCountBetween(startDate, endDate);
    }
}

File: src/main/kotlin/com/example/gava/domain/todo/controller/TodoController.kt
package com.example.gava.domain.todo.controller

import com.example.gava.domain.todo.dto.CreateSubTodoRequest
import com.example.gava.domain.todo.dto.CreateTodoRequest
import com.example.gava.domain.todo.dto.SubTodoResponse
import com.example.gava.domain.todo.dto.TodoResponse
import com.example.gava.domain.todo.service.TodoService
import com.example.gava.security.CustomUserDetails
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import java.time.LocalDate

@RestController
@RequestMapping("/api/todos")
class TodoController(private val todoService: TodoService) {
    @PostMapping
    fun create(
        @RequestBody createTodoRequest: CreateTodoRequest,
        @AuthenticationPrincipal userDetails: CustomUserDetails,
    ): ResponseEntity<TodoResponse> {
        val todoResponse = todoService.create(createTodoRequest, userDetails.getUserId())

        return ResponseEntity.ok(todoResponse)
    }

    @GetMapping
    fun getAll(
        @RequestParam(required = false) date: LocalDate?,
        @RequestParam(required = false) groupId: Long?,
        @RequestParam(required = false) done: Boolean?,
        @AuthenticationPrincipal userDetails: CustomUserDetails
    ): ResponseEntity<List<TodoResponse>> {
        val todoResponses = todoService.getAll(userDetails.getUserId(), date, groupId, done)

        return ResponseEntity.ok(todoResponses)
    }
    
    @PutMapping("sub-todo/{id}")
    fun updateSubTodo(
        @PathVariable id: Long,
        @RequestBody subTodoRequest: CreateSubTodoRequest,
        @AuthenticationPrincipal userDetails: CustomUserDetails
    ): ResponseEntity<SubTodoResponse> {
        val subTodoResponse = todoService.updateSubTodo(id, subTodoRequest, userDetails.getUserId())

        return ResponseEntity.ok(subTodoResponse)
    }

    @PatchMapping("/{id}/done")
    fun done(
        @PathVariable id: Long,
        @AuthenticationPrincipal userDetails: CustomUserDetails
    ): ResponseEntity<Void> {
        todoService.done(id, userDetails.getUserId())
        return ResponseEntity.noContent().build()
    }

    @DeleteMapping("/{id}")
    fun delete(
        @PathVariable id: Long,
        @AuthenticationPrincipal userDetails: CustomUserDetails
    ): ResponseEntity<Unit> {
        todoService.delete(id, userDetails.getUserId())

        return ResponseEntity.noContent().build()
    }
}
